File: packages/go-sdk/operations/rbac.go
package operations

import "context"

// RBACService provides role-based access control operations.
type RBACService struct{ r Requester }

func NewRBACService(r Requester) *RBACService { return &RBACService{r: r} }

func (s *RBACService) ListAuditEvents(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/audit-events", query, nil, nil, &result)
	return result, err
}

func (s *RBACService) ListMembers(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/members", query, nil, nil, &result)
	return result, err
}

func (s *RBACService) CreateMember(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/members", nil, nil, body, &result)
	return result, err
}

func (s *RBACService) DeleteMember(ctx context.Context, userID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/members/{userID}", map[string]string{"userID": userID}), nil, nil, nil, nil)
}

func (s *RBACService) BulkMember(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/members/bulk", nil, nil, body, &result)
	return result, err
}

func (s *RBACService) ListRoles(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/roles", query, nil, nil, &result)
	return result, err
}

func (s *RBACService) CreateRole(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/roles", nil, nil, body, &result)
	return result, err
}

func (s *RBACService) GetRole(ctx context.Context, roleID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/roles/{roleID}", map[string]string{"roleID": roleID}), nil, nil, nil, &result)
	return result, err
}

func (s *RBACService) UpdateRole(ctx context.Context, roleID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "PATCH", PathParams("/v1/roles/{roleID}", map[string]string{"roleID": roleID}), nil, nil, body, &result)
	return result, err
}

func (s *RBACService) DeleteRole(ctx context.Context, roleID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/roles/{roleID}", map[string]string{"roleID": roleID}), nil, nil, nil, nil)
}

func (s *RBACService) ListResourcePolicies(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/resource-policies", query, nil, nil, &result)
	return result, err
}

func (s *RBACService) CreateResourcePolicy(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/resource-policies", nil, nil, body, &result)
	return result, err
}

func (s *RBACService) DeleteResourcePolicy(ctx context.Context, policyID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/resource-policies/{policyID}", map[string]string{"policyID": policyID}), nil, nil, nil, nil)
}

func (s *RBACService) ListTagPolicies(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/tag-policies", query, nil, nil, &result)
	return result, err
}

func (s *RBACService) CreateTagPolicy(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/tag-policies", nil, nil, body, &result)
	return result, err
}

func (s *RBACService) DeleteTagPolicy(ctx context.Context, policyID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/tag-policies/{policyID}", map[string]string{"policyID": policyID}), nil, nil, nil, nil)
}

func (s *RBACService) SeedRoles(ctx context.Context) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/seed-roles", nil, nil, nil, &result)
	return result, err
}

File: packages/rust-sdk/src/operations/rbac.rs
use crate::client::StraitClient;
use crate::errors::StraitError;
use crate::http::substitute_path_params;
use std::sync::Arc;

pub struct RbacService {
    client: Arc<StraitClient>,
}

impl RbacService {
    pub fn new(client: Arc<StraitClient>) -> Self {
        Self { client }
    }

    pub async fn list_audit_events(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/rbac/audit-events", query, None, None)
            .await
    }

    pub async fn list_members(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/rbac/members", query, None, None)
            .await
    }

    pub async fn create_member(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/rbac/members", None, None, Some(body))
            .await
    }

    pub async fn delete_member(&self, user_id: &str) -> Result<(), StraitError> {
        let path = substitute_path_params("/v1/rbac/members/{userID}", &[("userID", user_id)]);
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }

    pub async fn bulk_member(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/rbac/members/bulk", None, None, Some(body))
            .await
    }

    pub async fn list_roles(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/rbac/roles", query, None, None)
            .await
    }

    pub async fn create_role(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/rbac/roles", None, None, Some(body))
            .await
    }

    pub async fn get_role(&self, role_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/rbac/roles/{roleID}", &[("roleID", role_id)]);
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn update_role(
        &self,
        role_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/rbac/roles/{roleID}", &[("roleID", role_id)]);
        self.client
            .do_request("PATCH", &path, None, None, Some(body))
            .await
    }

    pub async fn delete_role(&self, role_id: &str) -> Result<(), StraitError> {
        let path = substitute_path_params("/v1/rbac/roles/{roleID}", &[("roleID", role_id)]);
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }

    pub async fn list_resource_policies(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/rbac/resource-policies", query, None, None)
            .await
    }

    pub async fn create_resource_policy(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/rbac/resource-policies", None, None, Some(body))
            .await
    }

    pub async fn delete_resource_policy(&self, policy_id: &str) -> Result<(), StraitError> {
        let path = substitute_path_params(
            "/v1/rbac/resource-policies/{policyID}",
            &[("policyID", policy_id)],
        );
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }

    pub async fn list_tag_policies(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/rbac/tag-policies", query, None, None)
            .await
    }

    pub async fn create_tag_policy(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/rbac/tag-policies", None, None, Some(body))
            .await
    }

    pub async fn delete_tag_policy(&self, policy_id: &str) -> Result<(), StraitError> {
        let path = substitute_path_params(
            "/v1/rbac/tag-policies/{policyID}",
            &[("policyID", policy_id)],
        );
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }

    pub async fn seed_roles(&self) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/rbac/seed", None, None, None)
            .await
    }
}
